File: Linked_List_Cycle_II.cpp
/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @return: The node where the cycle begins. 
     *           if there is no cycle, return null
     */
    ListNode *detectCycle(ListNode *head) {
        // write your code here
        ListNode *dummy = new ListNode(INT_MIN);
        dummy->next = head;
        ListNode *fast = dummy, *slow = dummy;
        while (fast && fast->next){
            fast = fast->next->next;
            slow = slow->next;
            if (fast == slow){
                //check the circle
                fast = dummy;
                while (slow != fast){
                    slow = slow->next;
                    fast = fast->next;
                }
                return slow;
            }
        }
        return nullptr;
    }
};

File: Linked List Cycle II.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of the linked list.
    @return: The node where the cycle begins. 
                if there is no cycle, return null
    """
    def detectCycle(self, head):
        # write your code here
        
        # dummy = ListNode(-1)
        # dummy.next = head
        # fast, slow = dummy, dummy
        # while fast and fast.next:
        #     fast = fast.next.next
        #     slow = slow.next
        #     if slow == fast:
        #         break
        
        # if not fast or not fast.next:
        #     return None
        
        # slow = dummy
        # while slow != fast:
        #     slow = slow.next
        #     fast = fast.next
        # return slow

        #better codes
        if not head:
            return None

        fast, slow = head.next, head
        while fast != slow:
            if fast and fast.next:
                fast = fast.next.next
            else:
                return None
            slow = slow.next

        pre = head
        slow = slow.next
        while pre != slow:
            pre, slow = pre.next, slow.next
        return pre